src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/TableSchemaResolver.java

```diff
@@ -30,40 +32,72 @@ protected TableSchemaResolver(ColumnTypeMapper columnTypeMapper) {
    *
    * @param record the sink record that contains the schema and actual data
-   * @param columnNames the names of the extra columns
+   * @param columnsToInclude the names of the columns to include in the schema
    * @return a Map object where the key is column name and value is ColumnInfos
    */
-  public TableSchema resolveTableSchemaFromRecord(SinkRecord record, List<String> columnNames) {
-    if (columnNames == null) {
+  public TableSchema resolveTableSchemaFromRecord(
+      SinkRecord record, List<String> columnsToInclude) {
+    if (columnsToInclude == null || columnsToInclude.isEmpty()) {
       return new TableSchema(new HashMap<>());
     }
-    Map<String, ColumnInfos> columnToType = new HashMap<>();
-    Map<String, ColumnInfos> schemaMap = getSchemaMapFromRecord(record);
-    JsonNode recordNode = RecordService.convertToJson(record.valueSchema(), record.value(), true);
-    Set<String> columnNamesSet = new HashSet<>(columnNames);
-
-    Iterator<Map.Entry<String, JsonNode>> fields = recordNode.fields();
-    while (fields.hasNext()) {
-      Map.Entry<String, JsonNode> field = fields.next();
-      String colName = Utils.quoteNameIfNeeded(field.getKey());
-      if (columnNamesSet.contains(colName)) {
-        ColumnInfos columnInfos;
-        if (schemaMap.isEmpty()) {
-          // No schema associated with the record, we will try to infer it based on the data
-          columnInfos = new ColumnInfos(inferDataTypeFromJsonObject(field.getValue()), null);
-        } else {
-          // Get from the schema
-          columnInfos = schemaMap.get(field.getKey());
-          if (columnInfos == null) {
-            // only when the type of the value is unrecognizable for JAVA
-            throw SnowflakeErrors.ERROR_5022.getException(
-                "column: " + field.getKey() + " schemaMap: " + schemaMap);
-          }
-        }
-        columnToType.put(colName, columnInfos);
-      }
+
+    Set<String> columnNamesSet = new HashSet<>(columnsToInclude);
+
+    if (hasSchema(record)) {
+      return getTableSchemaFromRecordSchema(record, columnNamesSet);
+    } else {
+      return getTableSchemaFromJson(record, columnNamesSet);
     }
-    return new TableSchema(columnToType);
+  }
+
+  private boolean hasSchema(SinkRecord record) {
+    return record.valueSchema() != null
+        && record.valueSchema().fields() != null
+        && !record.valueSchema().fields().isEmpty();
+  }
+
+  private TableSchema getTableSchemaFromRecordSchema(
+      SinkRecord record, Set<String> columnNamesSet) {
+    JsonNode recordNode = RecordService.convertToJson(record.valueSchema(), record.value(), true);
+    Map<String, ColumnInfos> schemaMap = getFullSchemaMapFromRecord(record);
+    Map<String, ColumnInfos> columnsInferredFromSchema =
+        Streams.stream(recordNode.fields())
+            .map(ColumnValuePair::of)
+            .filter(pair -> columnNamesSet.contains(pair.getQuotedColumnName()))
+            .peek(
```

Comment:
I'm not a fan of `peek` as the docs explicitly stated: `This method exists mainly to support debugging, where you want to see the elements as they flow past a certain point in a pipeline:`
Maybe it would be better to partition/group it into separate streams?

Reply:
partitionBy looks ok